sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/Body.cs

```diff
@@ -162,5 +166,80 @@ private void Append(Data segment)
                 _segments.Add(memory.Slice(0, dataToAppend.Length));
             }
         }
+
+        /// <summary>
+        /// Eagerly copies the provided data segments into a single continuous buffer while still keeping around a list of the individual copied segments.
+        /// Important for the receive path in order to make sure the buffers managed by the underlying AMQP library can be released on dispose.
+        /// </summary>
+        private sealed class EagerCopyingBodyWithManagedBuffer : Body, IDisposable
```

Comment:
I wonder at what point the need for a contiguous block of memory becomes a challenge.   Do we think this may cause GC pressure in memory-constrained environments, or possibly force applications into smaller batch sizes than they'd otherwise prefer?

Reply:
I wonder if we would potentially want to offer this as an option or apply some kind of heuristic to consider patterns in batch and payload size averages against available memory.

Reply:
Not sure I understand. The concern about the block of memory is already today a problem. Isn't even today in a constraint environment the memory you require depending on the expected message sizes multiplied by the concurrency on the receiver? With the rent and return pattern this is no different no? The benefit though is that you are not allocated Gen0 garbage or with messages get bigger no LOH fragmentations which ultimately is better for constraint environments right?

Reply:
No argument on the same overall memory need, and I agree that the single allocation of the larger block is quite likely more efficient than paying for a greater number of Gen0 collections.

The part that I was thinking on was the contiguous nature of the shared buffer.  With the individual allocation, the memory manager has more freedom to grab smaller blocks without compacting free space.  I'm wondering if there are applications running on devices or in restricted container environments where finding a large block is possibly difficult and impactful.

I don't have anything fully concrete in mind.  Your rationale may very well be more correct.   I'm honestly not sure without thinking more deeply on it and potentially running scenarios to see what happens.